src/main/java/dev/mcnees/moneymapper/batch/MultilineQFXReader.java

```diff
@@ -0,0 +1,183 @@
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import dev.mcnees.moneymapper.domain.Transaction;
+
+import org.springframework.batch.item.ExecutionContext;
+import org.springframework.batch.item.ItemReader;
+import org.springframework.batch.item.ItemStream;
+import org.springframework.batch.item.ItemStreamException;
+import org.springframework.batch.item.NonTransientResourceException;
+import org.springframework.batch.item.ParseException;
+import org.springframework.batch.item.UnexpectedInputException;
+import org.springframework.batch.item.file.FlatFileItemReader;
+import org.springframework.batch.item.file.mapping.DefaultLineMapper;
+import org.springframework.batch.item.file.mapping.PassThroughFieldSetMapper;
+import org.springframework.batch.item.file.transform.DefaultFieldSetFactory;
+import org.springframework.batch.item.file.transform.FieldSet;
+import org.springframework.batch.item.file.transform.FieldSetFactory;
+import org.springframework.batch.item.file.transform.LineTokenizer;
+import org.springframework.batch.item.file.transform.PatternMatchingCompositeLineTokenizer;
+import org.springframework.core.io.Resource;
+
+
+public class MultilineQFXReader implements ItemReader<Transaction>, ItemStream {
+
+	private FlatFileItemReader<FieldSet> delegate;
+
+	private final StatementTransactionTokenizer statementTransactionTokenizer = new StatementTransactionTokenizer();
+
+	private static final DateTimeFormatter QFX_DATE_FORMAT = DateTimeFormatter.ofPattern("yyyyMMdd");
+
+	public MultilineQFXReader(Resource resource) {
+		delegate = new FlatFileItemReader<>();
+		
+		delegate.setResource(resource);
+
+		DefaultLineMapper lineMapper = new DefaultLineMapper();
+
+		PassThroughFieldSetMapper fieldSetMapper = new PassThroughFieldSetMapper();
+		lineMapper.setFieldSetMapper(fieldSetMapper);
+
+		lineMapper.setLineTokenizer(getPatternMatchingTokenizer());
+
+		delegate.setLineMapper(lineMapper);
+
+	}
+
+	private LineTokenizer getPatternMatchingTokenizer() {
+		PatternMatchingCompositeLineTokenizer lineTokenizer = new PatternMatchingCompositeLineTokenizer();
+
+		Map<String, LineTokenizer> tokenizerMap = new HashMap<>();
+
+		tokenizerMap.put("<STMTTRN>", statementTransactionTokenizer);
+		tokenizerMap.put("<TRNAMT>*", statementTransactionTokenizer);
+		tokenizerMap.put("<NAME>*", statementTransactionTokenizer);
+		tokenizerMap.put("<FITID>*", statementTransactionTokenizer);
+		tokenizerMap.put("<DTPOSTED>*", statementTransactionTokenizer);
+		tokenizerMap.put("</STMTTRN>", statementTransactionTokenizer);
+		tokenizerMap.put("*", new DefaultTokenizer()); //ignore anything that isn't a tag of interest
+
+		lineTokenizer.setTokenizers(tokenizerMap);
+
+		return lineTokenizer;
+	}
+
+	@Override
+	public Transaction read() throws Exception, UnexpectedInputException, ParseException, NonTransientResourceException {
```

Comment:
Why don't I want that?

Also there is a call to `this.delegate.read()` in the for loop a couple lines below.  Wouldn't that go through line by line?

I tried to follow the pattern in the [Batch documentation](https://docs.spring.io/spring-batch/docs/current/reference/html/common-patterns.html#multiLineRecords) which is also in the [Batch sample](https://github.com/spring-projects/spring-batch/blob/main/spring-batch-samples/src/test/java/org/springframework/batch/sample/iosample/internal/MultiLineTradeItemReader.java).

Reply:
It will go line by line, but what happens if something fails? In this example, the job would start over instead of picking up where it left off. I guess given the H2 aspect it's fine but not how most batch jobs would be done (as you wouldn't want to restart from the very beginning on a restart in most use cases).